docs/cookbook/custom-promotion-action.rst

```diff
@@ -0,0 +1,147 @@
+How to add a custom promotion action?
+=====================================
+
+Let's assume that you would like to have a promotion that gives **100% discount on the cheapest item in the cart**.
+
+See what steps need to be taken to achieve that:
+
+Create a new promotion action
+-----------------------------
+
+You will need a new class ``CheapestProductDiscountPromotionActionCommand``.
+
+It will give a discount equal to the unit price of the cheapest item. That's why it needs to have the Proportional Distributor and
+the Adjustments Applicator. The ``execute`` method applies the discount and distributes it properly on the totals.
+This class needs also a ``isConfigurationValid()`` method which was omitted in the snippet below.
+
+.. code-block:: php
+
+    <?php
+
+    namespace AppBundle\Promotion\Action;
+
+    class CheapestProductDiscountPromotionActionCommand extends DiscountPromotionActionCommand
+    {
+        const TYPE = 'cheapest_item_discount';
+
+        /**
+         * @var ProportionalIntegerDistributorInterface
+         */
+        private $proportionalDistributor;
+
+        /**
+         * @var UnitsPromotionAdjustmentsApplicatorInterface
+         */
+        private $unitsPromotionAdjustmentsApplicator;
+
+        /**
+         * @param ProportionalIntegerDistributorInterface $proportionalIntegerDistributor
+         * @param UnitsPromotionAdjustmentsApplicatorInterface $unitsPromotionAdjustmentsApplicator
+         */
+        public function __construct(
+            ProportionalIntegerDistributorInterface $proportionalIntegerDistributor,
+            UnitsPromotionAdjustmentsApplicatorInterface $unitsPromotionAdjustmentsApplicator
+        ) {
+            $this->proportionalDistributor = $proportionalIntegerDistributor;
+            $this->unitsPromotionAdjustmentsApplicator = $unitsPromotionAdjustmentsApplicator;
+        }
+
+        /**
+         * {@inheritdoc}
+         */
+        public function execute(PromotionSubjectInterface $subject, array $configuration, PromotionInterface $promotion)
+        {
+            if (!$subject instanceof OrderInterface) {
+                throw new UnexpectedTypeException($subject, OrderInterface::class);
+            }
+
+            $items = $subject->getItems();
+
+            $cheapestItem = $items->first();
+
+            $itemsTotals = [];
+
+            foreach ($items as $item) {
+                $itemsTotals[] = $item->getTotal();
+
+                $cheapestItem = ($item->getVariant()->getPrice() < $cheapestItem->getVariant()->getPrice()) ? $item : $cheapestItem;
+            }
```

Comment:
Yeah sure. But the constraints should be restricted by Rules. Not by the action itself :)